### DISCUSSION TITLE
copy barcodes of UAM:Ento to other id

### DISCUSSION THREAD

**Discussion by DerekSikes** [ArctosDB/arctos#4680] [General]
Issue Documentation is http://handbook.arctosdb.org/how_to/How-to-Use-Issues-in-Arctos.html

**Is your feature request related to a problem? Please describe.**

A few years back I had Dusty copy all the UAM:Ento specimen part barcodes to the 'other ID' field so these would be publicly visible.  The reason for this was to allow those with loans of specimens to use Arctos to find data on the specimens they had borrowed. Because barcodes on specimens are not publicly shared this was a problem and this was a solution to that problem.

However, a new related problem has arisen. GBIF, iDigBio, and all the data aggregators can act as a form of data-backup in cases of a crash & data loss. However, they don't have the barcodes of the specimens since these are not publicly shared. They DO have the other IDs, however. Thus, if one ever wants to use GBIF data to restore lost data it would be best if GBIF also had the part barcodes with each record.

** Describe what you're trying to accomplish** 
An clear and concise overview of the goals; why are you asking for this?

The past effort was a one-time script that copied part barcodes to the other id field. This was good but immediately became 'old' in that as new records were created they did not have barcodes copied into the other id field. Also, on rare occasions a specimen with a barcode value in the other ID field would get a new barcode leading to these two fields disagreeing. This can also happen from cloning records if the user forgets or doesn't know to edit the other ID field appropriately.

Thus, a better solution would be to have a script that UAM:Ento admin can run (or set to run on a schedule) that does the following:

1) copies all part barcodes to the other ID field if needed (if there isn't already a matching barcode in the other id field)
2) checks to see if barcodes in the other ID field match to those in the parts and if not, replaces them so they match.

**Describe the solution you'd like**
How might we accomplish your goals?

If this were a script that could be scheduled to run weekly that would be great.

**Describe alternatives you've considered**

Alternatively, can some collections choose to share their part & barcode data with GBIF et al. (are there even DWC fields to map to?) if so, this would be an easier solution - just let whomever chooses share these data publicly.

Personally, I'd feel a lot safer if all my collection's part barcode & other data were shared in all the aggregators so if Arctos vanishes or suffers some horrible crash I don't end up with 400k+ records with no barcode data!!

**Additional context**

**Priority**
Please assign a priority-label.

immensely high

**Comment by Jegelewicz** [comment]
> Alternatively, can some collections choose to share their part & barcode data with GBIF et al. (are there even DWC fields to map to?) if so, this would be an easier solution - just let whomever chooses share these data publicly.

This seems like the better option. Maybe the barcode could be built into the [material sample ID](https://dwc.tdwg.org/terms/#dwc:materialSampleID)?

**Comment by campmlc** [comment]
I agree absolutely with this issue as I've just lost a month of barcodes
and object tracking. Whatever method works most effectively.

**Comment by dustymc** [comment]
>on rare occasions a specimen with a barcode value in the other ID field would get a new barcode leading to these two fields disagreeing.

This seems like a very good thing to me - it's the only approach that can serve the stated objective of using barcodes to persistently identify records (rather than just find parts from current data). I suggest this should be additive, not a replacement.

>are there even DWC fields to map to

There could be (but "fields" not actual fields) - ArctosDB/documentation-wiki#258

>material sample ID

Seems more correct, I can map that, I'll probably need guidance, I'm not sure if GBIF accepts it or not.

Anything involving site-wide exposure of barcodes requires discussion - I think some collections see barcodes as "private."

I think the realistic immediate approach is a barcode-->otherID script that runs for a hard-coded list of collections. 

>Whatever method works most effectively.

@campmlc I think that's your call. Most insects have a single barcoded part, so having the value somewhere in the Occurrence is almost as good as anything else that could happen. Having 30 barcode values mixed in with the 30 non-barcode values in https://dwc.tdwg.org/terms/#dwc:otherCatalogNumbers, and 30 not-obviously-related parts in https://dwc.tdwg.org/terms/#dwc:preparations, is a different situation.

>immensely high

I am incredibly hesitant to do anything that might stress Oracle, and this would. That is hopefully 100% paranoia. Please advise, @ArctosDB/arctos-working-group-officers.

**Comment by Jegelewicz** [comment]
> I'm not sure if GBIF accepts it or not.

Even if they don't, the DwC file will be at VertNet as a backup, No?

> Anything involving site-wide exposure of barcodes requires discussion - I think some collections see barcodes as "private."

If they are embedded in a larger identifier, I think we can get everyone on board.

> barcode-->otherID
Holy crap, some MSB records will end up with 10 or more of these....plus, what @dusty said.

As with pretty much everything right now - I think this needs to wait a few weeks until PG migration.

**Comment by dustymc** [comment]
>DwC file will be at VertNet 

I believe that's correct; I think they version.

>MSB records will end up with 10 or more 

http://test.arctos.database.museum/guid/MSB:Mamm:292039 has 42

**Comment by campmlc** [comment]
Can they be a specific kind of other ID that can be displayed separately in
UI, like relationships?

**Comment by dustymc** [comment]
>specific kind of other ID 

Yes but no?

<img width="709" alt="Screen Shot 2020-05-14 at 5 17 19 PM" src="https://user-images.githubusercontent.com/5720791/81998209-cc503480-9606-11ea-93d3-c5ce82975858.png">

>displayed separately 

That sounds incredibly confusing.

**Comment by dustymc** [comment]
This is now possible via search-->part/table-->download-->otherID loader - is some other action necessary?

**Comment by Jegelewicz** [comment]
Candidate for a workflow as discussed in Object tracking today?

**Comment by campmlc** [comment]
This relates to idea of Other ID metadata and linking of Other IDs to parts and events for e.g. GenBank sequences derived from loaned parts. Ideally we could associate Other ID barcodes with their respective parts and solve the problem of the 42 barcodes (isn't that the answer to life, the universe, and everything?)

**Comment by Jegelewicz** [comment]
I was about to close this, but I am thinking that the "barcode" identifier should actually be a part attribute? This was not a thing when we were discussing this before, but it would eliminate the issue of 

>     MSB records will end up with 10 or more
> 
> http://test.arctos.database.museum/guid/MSB:Mamm:292039 has 42

thoughts? @DerekSikes @campmlc

**Comment by DerekSikes** [comment]
I don't understand, sorry.

-D

On Fri, May 13, 2022 at 2:06 PM Teresa Mayfield-Meyer <
***@***.***> wrote:

> I was about to close this, but I am thinking that the "barcode" identifier
> should actually be a part attribute? This was not a thing when we were
> discussing this before, but it would eliminate the issue of
>
> MSB records will end up with 10 or more
>
> http://test.arctos.database.museum/guid/MSB:Mamm:292039 has 42
>
> thoughts? @DerekSikes <https://github.com/DerekSikes> @campmlc
> <https://github.com/campmlc>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ArctosDB/arctos/issues/2685#issuecomment-1126543354>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACFNUMYGMRZ4PH3WSOQC2ELVJ3G5TANCNFSM4NA42DTA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

+++++++++++++++++++++++++++++++++++
*Derek S. Sikes*, Curator of Insects, Professor of Entomology
University of Alaska Museum (UAM), University of Alaska Fairbanks
1962 Yukon Drive, Fairbanks, AK   99775-6960
***@***.*** phone: 907-474-6278 he/him/his
University of Alaska Museum <https://www.uaf.edu/museum/collections/ento/>
-  search 357,704 digitized arthropod records
<http://arctos.database.museum/uam_ento>
+++++++++++++++++++++++++++++++++++

Interested in Alaskan Entomology? Join the Alaska Entomological
Society and / or sign up for the email listserv "Alaska Entomological
Network" at
http://www.akentsoc.org/contact_us

**Comment by Jegelewicz** [comment]
@DerekSikes the original request was to add the [barcode attribute](https://arctos.database.museum/info/ctDocumentation.cfm?table=ctcoll_other_id_type#barcode) to your records (great when you only have one part), but now we have [part attributes](https://arctos.database.museum/info/ctDocumentation.cfm?table=ctspecpart_attribute_type) and I was wondering if this would be a better part attribute so that records with 20 parts can match their barcodes with the appropriate part.

**Comment by DerekSikes** [comment]
Ah, I see.

The reason for adding barcodes to the attributes was to make them publicly
available so users who borrow specimens with barcodes on them can search
for them & see them etc. Will your new solution allow this?

-Derek

On Fri, May 13, 2022 at 2:15 PM Teresa Mayfield-Meyer <
***@***.***> wrote:

> @DerekSikes <https://github.com/DerekSikes> the original request was to
> add the barcode attribute to your records (great when you only have one
> part), but now we have part attributes and I was wondering if this would be
> a better part attribute so that records with 20 parts can match their
> barcodes with the appropriate part.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ArctosDB/arctos/issues/2685#issuecomment-1126547291>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACFNUM7UDFNLNJCUROA23GLVJ3H6VANCNFSM4NA42DTA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

+++++++++++++++++++++++++++++++++++
*Derek S. Sikes*, Curator of Insects, Professor of Entomology
University of Alaska Museum (UAM), University of Alaska Fairbanks
1962 Yukon Drive, Fairbanks, AK   99775-6960
***@***.*** phone: 907-474-6278 he/him/his
University of Alaska Museum <https://www.uaf.edu/museum/collections/ento/>
-  search 357,704 digitized arthropod records
<http://arctos.database.museum/uam_ento>
+++++++++++++++++++++++++++++++++++

Interested in Alaskan Entomology? Join the Alaska Entomological
Society and / or sign up for the email listserv "Alaska Entomological
Network" at
http://www.akentsoc.org/contact_us

**Comment by Jegelewicz** [comment]
> Will your new solution allow this?

Yep! It would be right here along with preservation
![image](https://user-images.githubusercontent.com/5725767/168399455-c8582b8a-1b13-4ea4-b2f0-34a7a6f4ca27.png)

**Comment by DerekSikes** [comment]
Then that sounds good.

Will people (the public without accounts) be able to search for these
barcodes using our search form?

-Derek


On Fri, May 13, 2022 at 3:00 PM Teresa Mayfield-Meyer <
***@***.***> wrote:

> Will your new solution allow this?
>
> Yep! It would be right here along with preservation
> [image: image]
> <https://user-images.githubusercontent.com/5725767/168399455-c8582b8a-1b13-4ea4-b2f0-34a7a6f4ca27.png>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ArctosDB/arctos/issues/2685#issuecomment-1126565039>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACFNUM4IH47RSI7CVPBWWXLVJ3NKDANCNFSM4NA42DTA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

+++++++++++++++++++++++++++++++++++
*Derek S. Sikes*, Curator of Insects, Professor of Entomology
University of Alaska Museum (UAM), University of Alaska Fairbanks
1962 Yukon Drive, Fairbanks, AK   99775-6960
***@***.*** phone: 907-474-6278 he/him/his
University of Alaska Museum <https://www.uaf.edu/museum/collections/ento/>
-  search 357,704 digitized arthropod records
<http://arctos.database.museum/uam_ento>
+++++++++++++++++++++++++++++++++++

Interested in Alaskan Entomology? Join the Alaska Entomological
Society and / or sign up for the email listserv "Alaska Entomological
Network" at
http://www.akentsoc.org/contact_us

**Comment by dustymc** [comment]
Yep, the "curatorial" search pane was made public a while back, anyone can search by barcode.

Persistent part identifiers is https://github.com/ArctosDB/arctos/issues/3630

Tentatively closing.

**Comment by Jegelewicz** [comment]
Nope - barcodes are not visible to public users
![image](https://user-images.githubusercontent.com/5725767/168403010-16abe143-dbc2-435a-80ce-41af4406c6ee.png)

So, should they be? Should we move the "barcode" attribute to be a part attribute? Or is everyone just happy with the way things are?

**Comment by campmlc** [comment]
Barcodes as part attributes make sense. That is public info. Might not be where most people would look.  For Derek, he can also just use the ID for a single barcode.
Maybe we need GenBank IDs as part attributes too.
